test/jdk/java/net/httpclient/http2/BadPushPromiseTest.java

```diff
@@ -0,0 +1,182 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8354276
+ * @library /test/lib /test/jdk/java/net/httpclient/lib
+ * @build jdk.test.lib.net.SimpleSSLContext jdk.httpclient.test.lib.http2.Http2TestServer
+ * @run testng/othervm
+ *      -Djdk.internal.httpclient.debug=true
+ *      -Djdk.httpclient.HttpClient.log=errors,requests,responses,trace
+ *      BadPushPromiseTest
+ */
+
+import jdk.httpclient.test.lib.common.HttpServerAdapters;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ProtocolException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpHeaders;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandlers;
+import java.net.http.HttpResponse.PushPromiseHandler;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static java.net.http.HttpClient.Version.HTTP_2;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.List.of;
+import static org.testng.Assert.*;
+
+public class BadPushPromiseTest {
+
+    private static final List<Map<String, List<String>>> BAD_HEADERS = of(
+            Map.of(":hello", of("GET")),                      // Unknown pseudo-header
+            Map.of("hell o", of("value")),                    // Space in the name
+            Map.of("hello", of("line1\r\n  line2\r\n")),      // Multiline value
+            Map.of("hello", of("DE" + ((char) 0x7F) + "L")),  // Bad byte in value
+            Map.of(":status", of("200"))                     // Response pseudo-header in request
+    );
+
+    static final String MAIN_RESPONSE_BODY = "the main response body";
+
+    HttpServerAdapters.HttpTestServer server;
+    URI uri;
+
+    @BeforeTest
+    public void setup() throws Exception {
+        server = HttpServerAdapters.HttpTestServer.create(HTTP_2);
+        HttpServerAdapters.HttpTestHandler handler = new ServerPushHandler(MAIN_RESPONSE_BODY);
+        server.addHandler(handler, "/");
+        server.start();
+        int port = server.getAddress().getPort();
+        System.err.println("Server listening on port " + port);
+        uri = new URI("http://localhost:" + port + "/foo/a/b/c");
```

Comment:
Hello Daniel, I think we should avoid `localhost` usage here and instead use `URIBuilder` test library util and then pass it  `InetAddress.getLoopbackAddress()` as the host. That should help avoid issues related to `localhost` being mapped to some other address and at the same time allow constructing the right URL when `InetAddress.getLoopbackAddress()` returns an IPv6 address.

Reply:
Ah - good point - you can also simply use `new URI(http://" + server.serverAuthority() + "/foo/a/b/c")`

Reply:
Good catch! Switched to `serverAuthority()`.